device-server/src/main/kotlin/com/badoo/automation/deviceserver/host/EmulatorsNode.kt

```diff
@@ -0,0 +1,318 @@
+                        state = DeviceState.FAILED,
+                        emulator = "",
+                        model = desiredCaps.model ?: "default-docker-snapshot",
+                        last_error = lastError
+                )
+                deallocateEmulatorSlot(emulator)
+            }
+        }
+    }
+
+    private fun httpGet(endpoint: String): Response =
+            httpClient.newCall(Request.Builder().url("http://$ticketHost/$endpoint").build()).execute()
+
+    private fun throwUnlessOK(it: Response) {
+        if (it.code() != HttpURLConnection.HTTP_OK) {
+            throw RuntimeException("NOT HTTP_OK: $it")
+        }
+    }
+
+    private fun releaseOnTicketMachine(deviceRef: DeviceRef): Boolean {
+        httpGet("docker/release/$deviceRef")
+                .use {
+                    throwUnlessOK(it)
+                    val body = it.body()?.string()
+                    val released = body?.contains(":SUCCESS: $deviceRef released") ?: false
+                    if (!released) {
+                        logger.info("Unable to release $deviceRef: $body $it")
+                    }
+                    return released
+                }
+    }
+
+    private fun emulatorFromRef(deviceRef: DeviceRef): EmulatorSerial =
+            (refToDto[deviceRef] ?: throw RuntimeException("No emulator for $deviceRef")).info.udid
+
+    private fun emulatorOrNullFromRef(deviceRef: DeviceRef): EmulatorSerial? {
+        return try {
+            emulatorFromRef(deviceRef)
+        } catch (e: RuntimeException) {
+            logger.warn(e.toString())
+            null
+        }
+    }
+
+    private fun updateDtoForRef(ref: String, state: DeviceState, emulator: EmulatorSerial,
+                                adb_socket: URI = URI(""), model: String,
+                                last_error: Throwable?): DeviceDTO {
+        val deviceDTO = DeviceDTO(
+                ref = ref,
+                state = state,
+                fbsimctl_endpoint = URI(ticketHost),
+                wda_endpoint = adb_socket,
+                info = DeviceInfo(udid = emulator, model = model, os = "", arch = "", name = ""),
+                calabash_port = 0, user_ports = HashSet(), capabilities = null, last_error = last_error?.toDto())
+        refToDto[ref] = deviceDTO
+        return deviceDTO
+    }
+
+    private fun deallocateEmulatorSlot(emulator: EmulatorSerial): DeviceDTO? {
+        return allocatedEmulators.remove(emulator)
+    }
+
+    private fun allocateEmulatorSlot(dto: DeviceDTO): EmulatorSerial {
+        synchronized(allocatedEmulators) {
+            try {
+                val emulator =
+                        (knownEmulators - allocatedEmulators.keys).asSequence().sorted().first()
+                allocatedEmulators[emulator] = dto
+                return emulator
+            } catch (e: RuntimeException) {
+                throw RuntimeException("No emulators left in $ticketHost pool", e)
```

Comment:
This is called before talking to the ticket machine, because we now tell ticket machine which emulator-XXXX we want. If someone manages to ask this node for emulators it doesn't have, we explode.